### PR TITLE
feat: codeowners file

### PR DESCRIPTION
**Pull request by Hunter-Thompson**
Gives the ability to add a CODEOWNERS file on the repo.

```
  codeOwners: {
    owners: [{
      owners: [
        'Hunter-Thompson',
      ],
      patterns: '*'
    }]
  }


project.github.codeOwners.addOwners({
  owners: [
    'Raoul-Duke',
  ],
  patterns: '*.js'
})
```

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by Hunter-Thompson** [comment]
Builds failing due to #805

**Review by eladb — COMMENTED** [review]
Wouldn't it makes sense to provide a higher level API than just an array of strings?

**Comment by hoegertn** [comment]
Additionally, the default value doe snot really make sense imho. So maybe it would be enough to have one field with the content (high-level) and then create the file or not. I do not see much value in a boolean activation flag and the content option.

**Comment by Hunter-Thompson** [comment]
> Wouldn't it makes sense to provide a higher level API than just an array of strings?

Could you provide an example? Il work on a Component based of the example.

> Additionally, the default value does not really make sense imho. So maybe it would be enough to have one field with the content (high-level) and then create the file or not. I do not see much value in a boolean activation flag and the content option.

Makes sense, will work on updating the PR.

**Comment by eladb** [comment]
@Hunter-Thompson when I say higher-level API I mean something like this:

```ts
const owners = new CodeOwners(project);
owners.addOwners('*', '@global-owner1', '@global-owner2');
owners.addOwners('*.js', '@js-owner');
owners.addOwners('*.go', 'docs@example.com');
owners.addOwners('/build/logs/', '@doctocat');
owners.addOwners('docs/*', 'docs@example.com');
// ...
```

Additionally, we can offer an initializer API:

```ts
const owners = new CodeOwners(project, {
  owners: [
    { pattern: '*', owners: [ '@global-owner1', '@global-owner2' ] },
    { pattern: '*.js', owners: [ '@js-owner' ] },
  ],
});

owners.addOwners(...);
```

**Review by eladb — CHANGES_REQUESTED** [review]
see comment

**Comment by Hunter-Thompson** [comment]
> @Hunter-Thompson lmk if you'd like to follow up here or should we close for now

@eladb, will pick this up over the weekend.

**Comment by Hunter-Thompson** [comment]
@eladb, made the required changes and updated the PR description. Let me know where you'd like changes.

File: src/github/codeowners.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+}
+export class CodeOwners extends Component {
+  private owners: string[] = [];
+
+  constructor(github: GitHub, options: CodeOwnersProps) {
+    super(github.project);
+    options.owners?.forEach(this.addOwners.bind(this));
+  }
+
+  public addOwners(owner: Owner) {
```
Comment (eladb) [inline src/github/codeowners.ts:22]: Why not?

```ts
addOwners(pattern: string, ...owners: string[])
```

File: src/github/codeowners.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import { GitHub } from './github';
+
+
+export interface Owner {
+  readonly patterns?: string;
+  readonly owners?: string[];
+}
+
+export interface CodeOwnersProps {
+  readonly owners?: Owner[];
```
Comment (eladb) [inline src/github/codeowners.ts:12]: Why not?

```ts
{
  owners: {
    pattern: [ owners ],
    ...
  }
  
}
```

**Comment by eladb** [comment]
@Hunter-Thompson could you please update the APIs to the proposed option. I think the ergonomics are better.

**Comment by Hunter-Thompson** [comment]
> @Hunter-Thompson could you please update the APIs to the proposed option. I think the ergonomics are better.

Sure thing, will make the changes when I get some time tomorrow.

**Comment by eladb** [comment]
?

**Comment by Hunter-Thompson** [comment]
> ?

Not getting time to work on this feature, please gimme a day or two.

**Pull request closed without merging**